base-containers/base/inginious_container_api/run_student.py

```diff
@@ -52,15 +53,12 @@ def run_student(cmd, container=None,
         - 253 means that the command timed out
         - 254 means that an error occurred while running the proxy
     """
-    if not os.path.exists("/.__input/__shared_kernel"):  # disable run_student when the kernel is not shared (TODO fix me)
-        print("run_student is not available with Kata yet")
-        return 251
-
-    elif run_as_root:  # Allowing root access to student is forbidden for now (TODO fiw me)
-        print("run_student as root is not available yet")
-        return 251
-
 
+    #  Checking runtimes
+    shared_kernel = os.path.exists("/.__input/__shared_kernel")
+    only_dockers = shared_kernel and not run_as_root
```

Comment:
shared_kernel and run_as_root should be mutually exclusive, and it should be (re)checked here I think

Reply:
Well actually, "shared_kernel" simply means that the container is running on docker. Damn it's true that it's not explicit ! I will rename this.
The "only_dockers" means that both grading_container and student_container are on docker.

So they are not excluse since we can have shared_kernel and only_dockers or not (docker-docker or kata-docker).
Actually the only situation to avoid is shared_kernel and run_as_root, still it's a nice idea to recheck this right there :)

Reply:
Nevermind, I missunderstood your comment, you are right shared_kernel and run_as_root are exclusive and should be checked here :)

Reply:
Still not clear for me... could you give us a logic table to describe what happens in each cases ?

Reply:
Here, we are in run_student.py so we are on the grading_container.
shared_kernel simply means that the current container (grading_container) is using a shared_kernel runtime (aka: docker runtime)
run_as_root is the argument passed by the user to run_student and represents the fact we want to start a student_container to run code in it as normal user or as root.

Thus there is no mutual exclusions, here are the different scenarios, remember we are in the grading_container so shared-kernel says nothing about the student_container:

shared_kernel  -  run_as_root            
True                     True                      Docker -> Kata
True                     True                      Docker -> Docker (aka usual scenario)
False                    True                     Kata -> Kata
False                    False                   Kata -> Docker

Reply:
I believe there is no issue here so I resolve the conversation.
The only dangerous scenario was about having a student_container with shared_kernel runtime and run_as_root together which is verified here: https://github.com/darkyne/INGInious/blob/60a915a8ab7fdad457777a2832ee110f44fdf8a1/base-containers/base/bin/_run_student_intern#L216

Don't hesitate if it's still not clear :)

Reply:
Still not ok for me.

`shared_kernel` indicates whether the kernel is shared or not; if it is the case, NOBODY should be allowed to run as root.
`run_as_root` indicates whether the parent container asked to run the student one as root. this can be compromised and should not be trusted. 

Please add a check to verify that, that could be something like:
```
if shared_kernel:
   run_as_root = False
```

And add comments everywhere explaining what are all the variables.

Reply:
I don't get why shared_kernel and run_as_root is incompatible.
As explain before, shared_kernel does not mean there is a shared_kernel between grading_container and student_container. 
Actually, shared_kernel in run_student means the grading_container is running on docker runtime (or Kata if False) and shared_kernel in run_student_intern means the student_cotnainer is running on docker runtime (or kata if False).

So the scenario where a docker grading_container wants to start a kata student_container, means at the run_student file we have shared_kernel and run_as_root both set to true.

Do you want me to rename "shared_kernel" into "run_on_docker" or something like that?